File: src/main/java/org/snust/numbertheory/공약수2436.java
package org.snust.numbertheory;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 공약수2436 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        long a = Long.parseLong(st.nextToken());
        long b = Long.parseLong(st.nextToken());

        long divide = b / a;
        long minSum = Long.MAX_VALUE;
        long answer = 0;
        long answer2=0;

        for(long i = 1; i < ((long) (Math.sqrt(divide)))+1; i++) {
            long j = divide /i;

            if (divide % i == 0 & gcd(i,j)==1){
                if (j-i < minSum){
                    minSum = j-i;
                    answer = i*a;
                    answer2 = j*a;
                }
            }
        }
        System.out.printf("%d %d%n",answer,answer2);
    }
    private static Long gcd(Long a, Long b) {
        while (a % b !=0) {
            long tmp = a%b;
            a = b;
            b = tmp;
        }
        return b;
    }
}

File: src/main/python/numbertheory/공약수2436.py
# 1. 갭을 줄여도 된다!
# 12 8
# 8 4

# 2. 최대 공약수라는 말은 두 수 중 하나로 나누어서 나머지가 생기지 않는다
# 8 4
# 최대공약수

# 3. 한 수를 가능한 만큼 갭을 줄인다
# 하나의 수를 작은수로 되는만큼 뺀다는 말은
# 나누고 나서의 나머지를 뜻한다.
#
# def _lcm(a,b):
#   return a*b//_gcd(a,b)
# --> lcm = a*b / gcd -> gcd * lcm = a*b

def _gcd(a,b):
  while a % b != 0:
    tmp = a % b
    a = b
    b = tmp
  return b

A, B = map(int, input().split())

divide = B // A # p와 q의 곱이어야 함
min_sum = 2000000000
answer = 0
# 두 수의 최대 공약수가 6이면 두 수 모두 6을 약수로 가지고 있다
# 최소 공배수는 서로소인 두 수의 곱과 최대공약수를 곱한 값이다

for i in range(1, int(divide**0.5)+1):
  j = int (divide/i)

  if divide % i == 0 and _gcd(i,j) == 1: # 나머지가 없고 서로소이면
    if j-i < min_sum:
      min_sum = j - i # 가장 작은 차이여야 두 수가 최소임
      answer = f"{i*A} {j*A}"

print(answer)

